[v15] Bugfixes  + spanish translation

pcastelovigo opened the pull request:
Glitch on product form with fleet_type_id when opened from rental order.

Commit:
corregir glitch

pcastelovigo committed:
traduccion castellano/spanish translation

olaf-wagner commented:
Hello [pcastelovigo](https://github.com/pcastelovigo),
thank you for the translations and the view fix. We will merge them locally at elego as the branches here are just synchronized from an internal repo. I'll ask @yweng8111 for a quick review and do the merge when he approves.
It may take some days till he is available though. Please stay tuned and thanks again.

pcastelovigo commented:
Glitch capture 
![glitch](https://github.com/elego/rental-vertical/assets/20586382/a2878b6b-0799-4ccf-a133-0498452084f9)

We were working with this for some time and we have some more code to releasse;
We plan to work with this vertical again and would like to clone from here, but feel free to merge or not next PRs
thank you for your great work

pcastelovigo committed:
rental pricelist by timeunit

yweng8111 commented on rental_product_variant/views/product_view.xml:
```diff
@@ -57,11 +57,9 @@
                     </button>
                 </div>
                 <xpath expr="//field[@name='categ_id']" position="replace" />
-                <xpath expr="//field[@name='type']" position="after">
```
to keep the origin requirement, you can change it into "detailed_type"

pcastelovigo replied:
done, much better

yweng8111 commented on rental_pricelist_timeunit/models/models.py:
```diff
@@ -0,0 +1,37 @@
+
+		product_context = dict(self.env.context, partner_id=self.order_id.partner_id.id, date=self.order_id.date_order, uom=self.product_uom.id)
+
+		final_price, rule_id = self.order_id.pricelist_id.with_context(product_context).get_product_price_rule(product or self.product_id, self.product_uom_qty or 1.0, self.order_id.partner_id)
```
Great job! Thanks for your contribution!
Can you please also do the improvement for the price without discount. Just calculate the final_price and base_price also in this way.
Can you please change the name of this file into sale_order_line.py and change the class name into SaleOrderLine

Can you please change the module name into rental_pricelist_timeunit_discount ?

pcastelovigo replied:
this was a client requirement. 
all requested changes done, watch out for unintended consquences with final_price and base_price changes, but everything seems to work OK

yweng8111 replied:
I've tested the feature "without discount". It doesn't work. The field "discount" of sale.order.line still depends on product_uom_qty. Maybe you should override the function _onchange_discount of sale.order.line

pcastelovigo replied:
I'll check this tomorrow, along with another bug I found with pricelists

pcastelovigo replied:
> I've tested the feature "without discount". It doesn't work. The field "discount" of sale.order.line still depends on product_uom_qty. Maybe you should override the function _onchange_discount of sale.order.line

[#f34edb7](https://github.com/elego/rental-vertical/pull/1/commits/f34edb78fff08e4b7bef36671ca24a3df1af58d8) makes this work and the same time fixes malfunction using regular pricelists

- Clarification: 
When you rented 2 items for 30 days it was like renting 60 days to the rent pricelist function. The new addon changes this behaviour and applies the price independently of amount of items. This was a client requirement.

Unrelated, regular pricelists and discount field was not working, override the function _onchange_discount of sale.order.line was the solution.

yweng8111 replied:
I think they belong to the same Feature. Can you please override the function _onchange_discount to make it also work for this Option. For me it is a bug if someone use the second Option and get a different price.
![Screenshot from 2023-11-22 13-41-53](https://github.com/elego/rental-vertical/assets/17932277/6a4d3251-01bd-45d3-8852-41aac9a23a91)

pcastelovigo replied:
Done, was done before, but duplicated

pcastelovigo committed:
keep the origin requirement

pcastelovigo committed:
improvement for the price without discount

pcastelovigo committed:
improvement for the price without discount

pcastelovigo committed:
fix bug UncaughtPromiseError

pcastelovigo commented:
to reproduce bug fixed in 4b0efe5

Go to stock > product > action_product_tmpl_forecast_report

yweng8111 commented:
> to reproduce bug fixed in [4b0efe5](https://github.com/elego/rental-vertical/commit/4b0efe5ad7f1e03cb63139decf57ef4094df803e)
> 
> Go to stock > product > action_product_tmpl_forecast_report

@vancouver29 Please check this commit.

pcastelovigo commented:
> > to reproduce bug fixed in [4b0efe5](https://github.com/elego/rental-vertical/commit/4b0efe5ad7f1e03cb63139decf57ef4094df803e)
> > Go to stock > product > action_product_tmpl_forecast_report
> 
> @vancouver29 Please check this commit.

![imagen](https://github.com/elego/rental-vertical/assets/20586382/33d0a3fa-8ade-4e5c-9fe1-e531647ae036)

pcastelovigo committed:
improvement + bugfix

yweng8111 commented on rental_pricelist/models/sale.py:
```diff
@@ -277,8 +276,18 @@ def product_uom_change(self):
                         )
                         break
+        if self.number_of_time_unit:
+            self.product_uom_qty = self.rental_qty * self.number_of_time_unit
```
It was already done in function rental_qty_number_of_days_change()

pcastelovigo replied:
Without that line, update_prices method calculates incorrect product_uom_qty

yweng8111 replied:
rental_qty_number_of_days_change changes product_uom_qty, so that the other onchange function will be triggered again with the correct product_uom_qty.

pcastelovigo replied:
This was correct, I deleted those lines in next commit

yweng8111 commented on rental_pricelist/models/sale.py:
```diff
@@ -277,8 +276,18 @@ def product_uom_change(self):
+    @api.onchange(
+        "product_id", "price_unit", "product_uom", "product_uom_qty", "tax_id"
+    )
+    def _onchange_discount(self):
```
It was already done in function rental_qty_number_of_days_change()

pcastelovigo replied:
Without that line, update_prices method calculates incorrect product_uom_qty

Edit:
Steps to reproduce: Make a rental quotation 2 items for 5 time units. Price correctly set as <2 * 5 * Price>
Choose another pricelist and run update_prices method
Price updates as <2 * price> missing the time units


update_price calls both **product_uom_change** and **_onchange_discount**

yweng8111 replied:
I've just installed module rental_pricelist and follow these steps I can not reproduce the error. if you change the pricelist, It doesn't reset the product_uom_qty in sale.order.line. So its price keeps unchanged. Did you installed some other modules, that change the product_uom_qty?

pcastelovigo replied:
In rental quotations, if you select amount of time, amount of units and change pricelists it resets product_uom_qty

I can make a video if still not sure

yweng8111 commented on rental_product_variant/views/product_view.xml:
```diff
@@ -57,11 +57,12 @@
                     <field name="categ_id" />
                 </xpath>
                 <xpath expr="//field[@name='barcode']" position="before">
+                    <field name="categ_id" />
```
Duplicated line

pcastelovigo replied:
[correct](https://github.com/elego/rental-vertical/pull/1/commits/c59e7969a94f63e487e1c42728edd5cd713f62c6)

pcastelovigo committed:
duplicated line

pcastelovigo committed:
delete unnecesary calc

pcastelovigo commented:
Just because this is getting too large:

- Glitch fully fixed. Duplicated line found by @yweng8111 deleted
 
- Additional addon commited didnt need changes. Reverted to original code

- Bug with pricelists was the same as discount field feature request. **_onchange_discount** function overrided via super() method. Unnecesary override found by @yweng8111 was also deleted.

Waiting for @vancouver29 to review the exception in Forecast button and I think this becomes free of bugs

pcastelovigo committed:
proper text formatting
